### PR TITLE
Non kube jobs

### LINKED ISSUE #83: Make `Job` generic module

### PR DESCRIPTION
**Pull request by smondet**
So cf. #82, it fixes #83, #84, #86.

I left this for a later PR: https://github.com/hammerlab/coclobas/issues/89

### CODE REVIEW

File: src/lib/job.ml [hunk 1]
```diff
@@ -0,0 +1,127 @@
+open Internal_pervasives
+
+module Specification = struct
+  type t =
+    | Kube of Kube_job.Specification.t
+    | Local_docker of Local_docker_job.Specification.t
+  [@@deriving yojson, show]
+  let kind =
```
Comment (ihodes) [inline src/lib/job.ml]: What's the purpose of this (just curious)?

File: src/ketrew_backend/plugin.ml [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
 
 module Run_parameters = struct
 
   type created = {
     client: Client.t;
-    specification: Kube_job.Specification.t [@main];
-    program: Ketrew_pure.Program.t option; (* Kept for display purposes *)
+    specification: Job.Specification.t [@main];
+    program: Ketrew_pure.Program.t option;
+    playground_path: string option;
```
Comment (smondet) [inline src/ketrew_backend/plugin.ml]: Playground goes by "ketrew"'s definition: a directory to put scripts and tmp-files.

There is an exception only in the case where you ask of a `` `Local_docker``
job, **with** a `Program.t`, and no playground.
All other cases are fine without a playground
(`` `Local_docker`` + one command; Kubernetes + `Program.t`; … etc.)

File: src/lib/job.ml (hunk 1 above)
Comment (smondet) [inline src/lib/job.ml]: It's just a way of pattern matching on the kind of Job without exposing the contents of `t`.
I ended up exposing `t` as a `private` type anyway (I think mostly for the command **display** in the Ketrew plugin) but the goal would be to make it completely abstract at some point.

File: src/ketrew_backend/plugin.ml (hunk 2 above)
Comment (ihodes) [inline src/ketrew_backend/plugin.ml]: Could we call it "temporary_directory" instead?

Comment (smondet) [inline src/ketrew_backend/plugin.ml]: I prefer sticking to Ketrew's wording so that all plugins look more similar.

**Comment by iskandr** [comment]
For the uninitiated, can you explain what this PR does?

**Comment by smondet** [comment]
@iskandr apart from some refactoring to abstract jobs and clusters, it adds anew kind of job: `` `Local_docker`` which is for running jobs locally (with `docker run ... -d`).
The new “local” kind of cluster inherits the pre-scheduling that we were doing to avoid DoSing kubernetes; so the local jobs are queued when we reach a given max-number of jobs.

**Pull request merged**